### ISSUE TITLE
Invalid environment settings

### ISSUE THREAD

**Issue by Arraying** [mauricioaniche/ck#92]
Sometimes, running CK will cause:
```
Exception in thread "main" java.lang.IllegalStateException: invalid environment settings
        at org.eclipse.jdt.core.dom.ASTParser.getClasspath(ASTParser.java:259)
        at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:1001)
        at com.github.mauricioaniche.ck.CK.calculate(CK.java:113)
        at com.github.mauricioaniche.ck.CK.calculate(CK.java:62)
        at com.github.mauricioaniche.ck.Runner.main(Runner.java:43)
```

For the cause of this, see [this reply](https://github.com/mauricioaniche/ck/issues/92#issuecomment-1010321127)

<details>
<summary>Original Issue, hidden by default</summary>

Sometimes, running CK will cause:
```
Exception in thread "main" java.lang.IllegalStateException: invalid environment settings
        at org.eclipse.jdt.core.dom.ASTParser.getClasspath(ASTParser.java:259)
        at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:1001)
        at com.github.mauricioaniche.ck.CK.calculate(CK.java:113)
        at com.github.mauricioaniche.ck.CK.calculate(CK.java:62)
        at com.github.mauricioaniche.ck.Runner.main(Runner.java:43)
```

This occurs arbitrarily, it was working on my machine, now it does not. A groupmate tried the exact same script on his laptop, where it worked, but it did not work on his desktop. 

We run the following inside our `repository/metrics` folder:
```sh
java -jar ck.jar ../ false 0 false ./output/
```
This is run on `java version "11.0.6" 2020-01-14 LTS`. 

Inside of `CK#calculate(Path, CKNotifier, Path...)` the paths to the .java files are all correct. There are no files there that should not be there. 

Curiously enough,  `ASTParser` is not 1000+ lines so the stacktrace is hard to follow. The method looks as such:
```java
        Main main = new Main(new PrintWriter(System.out), new PrintWriter(System.err), false, (Map)null, (CompilationProgress)null);
        ArrayList allClasspaths = new ArrayList();

        try {
            if ((this.bits & 32) != 0) {
                Util.collectRunningVMBootclasspath(allClasspaths);
            }

            int i;
            int max;
            if (this.sourcepaths != null) {
                i = 0;

                for(max = this.sourcepaths.length; i < max; ++i) {
                    String encoding = this.sourcepathsEncodings == null ? null : this.sourcepathsEncodings[i];
                    main.processPathEntries(4, allClasspaths, this.sourcepaths[i], encoding, true, false);
                }
            }

            if (this.classpaths != null) {
                i = 0;

                for(max = this.classpaths.length; i < max; ++i) {
                    main.processPathEntries(4, allClasspaths, this.classpaths[i], (String)null, false, false);
                }
            }

            ArrayList pendingErrors = main.pendingErrors;
            if (pendingErrors != null && pendingErrors.size() != 0) {
                throw new IllegalStateException("invalid environment settings");
            } else {
                return allClasspaths;
            }
        } catch (IllegalArgumentException var6) {
            throw new IllegalStateException("invalid environment settings", var6);
        }
```

It seems that `pendingErrors` is causing the problem, but the exception is supremely unhelpful and I have no idea what is going wrong. I have tried updating JDT but I cannot even find where it is released, their JavaDoc for the version I need is as good as non-existent. 

If anyone has faced this problem before and knows how to resolve it, please let me know. Otherwise, I will continue to look into it.

</details>

**Comment by Arraying** [comment]
I've run the method in question in isolation and it turns out it breaks due to the `.idea` folder. This information was contained in `pendingErrors`. Indeed, if I delete the folder everything runs as expected. Still not sure how this is caused

**Comment by Arraying** [comment]
It seems that this is because I use my project root directory as the `path` of CK, so it takes into account `.git` and `.idea` along with all the other files (I have a clean Gradle build folder at the moment, I do not want to imagine what it is like if it is not cleaned).

The specific error is:
```java
incorrect classpath: C:\Users\Paul\Programming\TUDelft\SEM\metrics\..\.idea\shelf\Uncommitted_changes_before_Update_at_09_01_2022_17_29_]
```
It seems like JDT performs some sort of truncation internally. I have no idea why exactly this fails. I've tried many, many things for hours but I could not get it to work. The only thing that worked is ignoring `.idea` folders like `.git` folders.

I propose the following steps:
1. Construct a list of ignored directories and initialize it with `.git`
2. Make a command line argument for additional ignored directory names.
3. Check against these ignored names (simple contains check) in `FileUtils`

Also, the way for checking if it is a Git directory does not work on Windows. In the above steps I will add a way to make it work on different operating systems. Also, currently paths can contain `..\`, I can quickly add a `.normalize()` to every `Path` so this does not happen.

What do you think, @mauricioaniche?

**Comment by mauricioaniche** [comment]
Errors happening inside the JDT code are usually due to some weird code that we throw at it. 

We should indeed ignore hidden directories. We can do that in the Utils class, as you suggested. A flag to allow users to exclude other directories can also be a good feature.

Would you mind implementing it ?

**Comment by Arraying** [comment]
I will get to this the coming days!

**Comment by mauricioaniche** [comment]
Merged already, let's close this issue!
